csharp/ql/test/library-tests/dataflow/modulusanalysis/ModulusAnalysis.cs

```diff
@@ -0,0 +1,63 @@
+using System;
+using System.Linq;
+
+class ModulusAnalysis
+{
+    const int c1 = 42;
+    const int c2 = 43;
+
+    void M(int i, bool cond, int x, int y, int[] arr, int otherSeven)
+    {
+        var eq = i + 3;
+
+        var mul = eq * c1 + 3; // congruent 3 mod 42
+
+        int seven = 7;
+        if (mul % c2 == seven)
+        {
+            System.Console.WriteLine(mul); // congruent 7 mod 43, 3 mod 42
+        }
+
+        if (otherSeven == 7)
+        {
+            if (mul % c2 == otherSeven)
+            {
+                System.Console.WriteLine(mul); // congruent 3 mod 42, 7 mod 43 missing
+            }
+        }
+
+        var j = cond
+            ? i * 4 + 3
+            : i * 8 + 7;
+        System.Console.WriteLine(j); // congruent 3 mod 4
+
+        if (x % c1 == 3 && y % c1 == 7)
+        {
+            System.Console.WriteLine(x + y); // congruent 10 mod 42
+        }
+
+        if (x % c1 == 3 && y % c1 == 7)
+        {
+            System.Console.WriteLine(x - y); // congruent 38 mod 42
+        }
+
+        var l = arr.Length * 4 - 11;
+        System.Console.WriteLine(l);
+
+        if (cond)
+        {
+            j = i * 4 + 3;
+        }
+        else
+        {
+            j = i * 8 + 7;
+        }
+        System.Console.WriteLine(j); // congruent 3 mod 4 or 7 mod 8
```

Comment:
Why is `7 mod 8` valid? If `cond = true` and `i = 0` then `j = 3` which is not `7 mod 8`.

Reply:
It's either of the two. I added an extra `if` after this line to check whether in a split control flow case we compute the correct modulus, and we do:
```
if (cond)
{
    j = i * 4 + 3;
}
else
{
    j = i * 8 + 7;
}
System.Console.WriteLine(j); // congruent 3 mod 4 or 7 mod 8
if (cond)
{
    System.Console.WriteLine(j); // congruent 3 mod 4
}
else
{
    System.Console.WriteLine(j); // congruent 7 mod 8
}
```

Also note that this is different from the `?:` operator, which we handle explicitly, without `phi` nodes:
```
var j = cond
    ? i * 4 + 3
    : i * 8 + 7;
System.Console.WriteLine(j); // congruent 3 mod 4
```
I'm not sure if this is a good approach though. The `?:` could be handled similarly to the `if`.

Reply:
I think this is wrong; I would expect `exprModulus` to be conjunctive and not disjunctive, isn't that the intention @aschackmull ? That is, I would expect the same result as in the `?:` case. I wonder if the same difference exists in Java?

Reply:
I pushed the same test case for Java in https://github.com/github/codeql/pull/4349/commits/e94d286a332761b941ea71e9b2854c0ce7d482ff, and there the result is what you expect.

Reply:
Isn't this the expected behaviour with control flow splitting?

Reply:
> Isn't this the expected behaviour with control flow splitting?

In some sense, yes. The problem is that we cannot tell whether two `exprModulus` results for the same expression means that there are two CFG nodes with each of the results, or whether both results hold for a single CFG node. I think the right thing to do (sorry for first realizing this now), is to use `ControlFlow::Node` instead of `Expr` throughout for C#.

Reply:
The same goes for sign analysis. Let us perhaps ignore splitting for now, and then I can try to redefine sign analysis in terms of CFG nodes instead of syntax nodes.